lib/auth/join/boundkeypair/boundkeypair.go

```diff
@@ -20,44 +20,69 @@ package boundkeypair
 
 import (
 	"context"
+	"crypto"
+	"encoding/json"
+	"log/slog"
 	"os"
 	"path/filepath"
+	"sync"
+	"time"
 
 	"github.com/gravitational/trace"
 	"golang.org/x/crypto/ssh"
 
 	"github.com/gravitational/teleport/api/utils/keys"
 	"github.com/gravitational/teleport/lib/auth/join"
 	"github.com/gravitational/teleport/lib/cryptosuites"
+	"github.com/gravitational/teleport/lib/sshutils"
 )
 
 const (
-	PrivateKeyPath = "id_bkp"
-	PublicKeyPath  = PrivateKeyPath + ".pub"
-	JoinStatePath  = "bkp_state"
+	PrivateKeyPath   = "id_bkp"
+	PublicKeyPath    = PrivateKeyPath + ".pub"
+	JoinStatePath    = "bkp_state"
+	KeyHistoryPath   = "bkp_key_history.json"
```

Comment:
Curious for feedback here, storing keys in a JSON list feels a _bit_ evil but was a fairly straightforward way to solve the problem. The active key is otherwise stored in plaintext in the same directory, so I don't really expect this to have any new security implications or anything like that.

(I also expect this to change if/when we add hardware key storage)

Reply:
I think JSON is fine too FWIW, but you _could_ also PEM-encode them and use block headers for the timestamp and any other metadata:

```Go
pem.Encode(out, &pem.Block{
	Type:    "PRIVATE KEY",
	Headers: map[string]string{"time": time.Now().Add(-5 * time.Minute).Format(time.RFC3339)},
	Bytes:   []byte(`key 1`),
})

pem.Encode(out, &pem.Block{
	Type:    "PRIVATE KEY",
	Headers: map[string]string{"time": time.Now().Format(time.RFC3339)},
	Bytes:   []byte(`key 2`),
})
```

```
-----BEGIN PRIVATE KEY-----
time: 2009-11-10T22:55:00Z

a2V5IDE=
-----END PRIVATE KEY-----
-----BEGIN PRIVATE KEY-----
time: 2009-11-10T23:00:00Z

a2V5IDI=
-----END PRIVATE KEY-----
```

I'm not sure if that really gets you much, though, and makes decoding more work.

Reply:
Fair enough, I'll keep it as is then, thanks for the feedback!

Reply:
> Perhaps worth having a top-level container struct that can hold the slice within?

I've added a new struct to wrap the entries. It now looks like:
```json
{
  "entries": [
    {
      "time": "2025-06-05T21:05:53.118742819-06:00",
      "private_key": "-----BEGIN PRIVATE KEY-----\nasdf\n-----END PRIVATE KEY-----\n"
    }
}
```